llvm/lib/Target/Mips/MipsAsmPrinter.cpp

```diff
@@ -565,12 +565,27 @@ bool MipsAsmPrinter::PrintAsmOperand(const MachineInstr *MI, unsigned OpNum,
       }
       break;
     }
-    case 'w':
-      // Print MSA registers for the 'f' constraint
-      // In LLVM, the 'w' modifier doesn't need to do anything.
-      // We can just call printOperand as normal.
+    case 'w': {
+      // Support to use MSA instructions on FPR.
+      // double a, b, c;
+      // asm volatile ("fmadd.d %w0, %w1, %w2" : "+f"(a) : "f"(b), "f"(c))
```

Comment:
I don't see what this C snippet adds?

Reply:
Without this patch, `asm volatile ("fmadd.d %w0, %w1, %w2" : "+f"(a) : "f"(b), "f"(c))` will emit
```
fmadd.d $f12, $f13, $f14"
```
And then the assembler rejects it.

Reply:
To use MSA on normal double or single.
Such as MIPSr5 has no `maddf.fmt`, while it may has MSA. So that we can use `fmadd.fmt` to implement `fma(3)` or `fmaf(3)`.

Reply:
Clearly if you don't implement a feature that feature doesn't work. My point is that you don't need a C snippet to illustrate what is going on here, especially if you simplify the code as I suggested. Otherwise the backend would be full of random bits of C in comments.